mass_mailing_partner/models/res_partner.py

```diff
@@ -0,0 +1,87 @@
+# Copyright 2015 Pedro M. Baeza <pedro.baeza@tecnativa.com>
+# Copyright 2015 Antonio Espinosa <antonio.espinosa@tecnativa.com>
+# Copyright 2015 Javier Iniesta <javieria@antiun.com>
+# Copyright 2017 David Vidal <david.vidal@tecnativa.com>
+# License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl.html).
+
+from odoo import _, api, fields, models
+from odoo.exceptions import ValidationError
+
+
+class ResPartner(models.Model):
+    _inherit = 'res.partner'
+
+    mass_mailing_contact_ids = fields.One2many(
+        string="Mailing contacts",
+        oldname="mass_mailing_contacts",
+        domain=[('opt_out', '=', False)],
+        comodel_name='mail.mass_mailing.contact', inverse_name='partner_id')
+    mass_mailing_contacts_count = fields.Integer(
+        string='Mailing contacts number',
+        compute='_compute_mass_mailing_contacts_count', store=True,
+        compute_sudo=True)
+    mass_mailing_stats = fields.One2many(
+        string="Mass mailing stats",
+        comodel_name='mail.mail.statistics', inverse_name='partner_id')
+    mass_mailing_stats_count = fields.Integer(
+        string='Mass mailing stats number',
+        compute='_compute_mass_mailing_stats_count', store=True)
+
+    @api.constrains('email')
+    def _check_email_mass_mailing_contacts(self):
+        for partner in self:
```

Comment:
Try just:

```py
if self.sudo().filtered("mass_mailing_contact_ids.list_ids"):
    raise ...
```

Reply:
I'm thinking that it's best to not limiting this, as it will lead to constant problems

Reply:
email field of contact is required and it depend on the email of the partner associated. So, i think, we should not permit that someone delete the email of the partner that has some contact associated.